### PR TITLE
[#2037] fix(mysql-catalog): Fix MySQL/PG wild card and case sensitive issues in table name

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Changes table name matching when query metadata of MySQL table.

### Why are the changes needed?

We need to consider some corner cases like wildcard character and case sensitivity issues. 

Fix: #2037 

### Does this PR introduce _any_ user-facing change?

N/A.

### How was this patch tested?

Add some test cases in `testMySQLTableCreatedByTrino` and test `testMySQLTableNameCaseSensitive`.

### CODE REVIEW

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 1]
```diff
@@ -402,4 +422,14 @@ protected JdbcColumn.Builder getBasicJdbcColumnInfo(ResultSet column) throws SQL
         .withNullable(nullable)
         .withDefaultValue(defaultValue);
   }
+
+  // Different JDBC catalog may have different escape character, please overwrite this method if
+  // necessary
+  protected String escapeSQL(String s) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:428]: escapeSQLIdentifier?

Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:428]: I afraid this is not called an `identifier`, like `where a = 'xxxxx'`, `xxxxx` is not an identifier strictly.

Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:428]: Do you have any other ideas about it? 
@diqiu50 @mchades

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 2]
```diff
@@ -117,18 +118,37 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
...
       // 1.Get table information
-      ResultSet table = getTable(connection, databaseName, tableName);
-      if (!table.next() || !tableName.equals(table.getString("TABLE_NAME"))) {
+      ResultSet table = getTable(connection, escapeSQL(databaseName), escapeSQL(tableName));
+      // The result of tables may be more than one due to the reason above, so we need to check the
+      // result
+      JdbcTable.Builder jdbcTableBuilder = new JdbcTable.Builder();
+      boolean found = false;
+      while (table.next()) {
+        if (Objects.equals(table.getString("TABLE_NAME"), tableName)) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:135]: is it possible, I create with a DB with name `A`, the underlying RDMS trans the name to `a`?  should we consider this?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 3]
```diff
@@ -117,18 +118,37 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
...
+          jdbcTableBuilder.withName(tableName);
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:136]: why not use getBasicJdbcTableInfo?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 2 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:135]: It's possible, but I'm not going to put the logic in the base class. Any JDBC catalog that is case-insensitive in the table name can overwrite this method, AFAIK, most JDBC tables are case-sensitive here.

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 4]
```diff
@@ -117,18 +118,37 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
 
   @Override
   public JdbcTable load(String databaseName, String tableName) throws NoSuchTableException {
+    // We should handle case sensitivity and wild card issue in some catalog tables, take a
+    // MySQL table for example.
+    // 1. MySQL will get table 'a_b' and 'A_B' when we query 'a_b' in a case-insensitive charset
+    // like utf8mb4.
+    // 2. MySQL will view 'a_b' as a wild card, and it will match any table name that starts with
+    // 'a',then any character and then 'b'.
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:126]: FYI:
"MySQL treats 'a_b' as a wildcard, matching any table name that begins with 'a', followed by any character, and ending with 'b'. "

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 5]
```diff
@@ -117,18 +118,37 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
 
   @Override
   public JdbcTable load(String databaseName, String tableName) throws NoSuchTableException {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:120]: besides loadTable, should we process the other DDL?  createTable, dropTable, AlterTable, etc

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 3 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:136]: Good, I will change it.

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:120]: I have checked it, others did not use `DatabaseMetaData`, so it's ok, you can check it.

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 6]
```diff
@@ -117,18 +118,37 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
...
+    // We should handle case sensitivity and wild card issue in some catalog tables, take a
+    // MySQL table for example.
+    // 1. MySQL will get table 'a_b' and 'A_B' when we query 'a_b' in a case-insensitive charset
+    // like utf8mb4.
+    // 2. MySQL will view 'a_b' as a wild card, and it will match any table name that starts with
+    // 'a',then any character and then 'b'.
     try (Connection connection = getConnection(databaseName)) {
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:127]: Don't you need to escape `databaseName` here?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 4 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:126]: fixed

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java [hunk 7]
```diff
@@ -1072,4 +1072,58 @@ public void testBackQuoteTable() {
...
+                tableCatalog.createTable(
+                    tableIdentifier2,
+                    newColumns,
+                    table_comment,
+                    properties,
+                    Transforms.EMPTY_TRANSFORM,
+                    Distributions.NONE,
+                    new SortOrder[0],
+                    indexes));
+    Assertions.assertEquals("TABLENAME", tableAgain.name());
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java:1127]: shoud we test loadTable?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 6 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:127]: Here is the exact match. Only the method `getTables` is fuzzy matching.
image

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java [hunk 8]
```diff
@@ -1072,4 +1072,58 @@ public void testBackQuoteTable() {
 
     Assertions.assertDoesNotThrow(() -> tableCatalog.dropTable(tableIdentifier));
   }
+
+  @Test
+  void testMySQLTableNameCaseSensitive() {
+    Column col1 = Column.of("col_1", Types.LongType.get(), "id", false, false, null);
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java:1078]: why mysql test NameCaseSensitive while pg test specialTableName? I think Mysql and PG need to test both cases.

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 1 above)
Comment (mchades) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:425]: `replaceWildcards` ?

Comment (mchades) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:425]: Whether the table name will contain dots `.`？

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 9]
```diff
@@ -117,18 +118,34 @@ public List<String> listTables(String databaseName) throws NoSuchSchemaException
...
+    // table for example.
+    // 1. MySQL will get table 'a_b' and 'A_B' when we query 'a_b' in a case-insensitive charset
+    // like utf8mb4.
+    // 2. MySQL treats 'a_b' as a wildcard, matching any table name that begins with 'a', followed
+    // by any character, and ending with 'b'.
     try (Connection connection = getConnection(databaseName)) {
       // 1.Get table information
-      ResultSet table = getTable(connection, databaseName, tableName);
-      if (!table.next() || !tableName.equals(table.getString("TABLE_NAME"))) {
+      ResultSet table = getTable(connection, escapeSQL(databaseName), escapeSQL(tableName));
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:129]: Whether the table name will contain dots `.`？

**Review by diqiu50 — COMMENTED** [review]
LGTM

**Review by FANNG1 — APPROVED** [review]

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java (hunk 1 above)
Comment (justinmclean) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java]: "Different JDBC catalogs may have different escape characters, please override this method if required"

Comment (yuqi1129) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java]: This method has been removed.
